subprojects/core-api/src/main/java/org/gradle/api/attributes/Attribute.java

```diff
@@ -67,7 +67,7 @@ private Attribute(String name, Class<T> type) {
         this.name = name;
         this.type = type;
         int hashCode = name.hashCode();
-        hashCode = 31 * hashCode + type.hashCode();
+        hashCode = 31 * hashCode + type.toString().hashCode();
```

Comment:
`Class#toString` allocates heap, so `type.getName().hashCode()` would probably be a better candidate for a hashcode instead of `toString`

```suggestion
        hashCode = 31 * hashCode + type.getName().hashCode();
```